src/Raven.Server/Documents/AI/Embeddings/QueryEmbeddingsCacher.cs

```diff
@@ -0,0 +1,113 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Raven.Server.Background;
+using Raven.Server.Documents.TransactionMerger.Commands;
+using Raven.Server.ServerWide.Context;
+using Sparrow.Server;
+
+namespace Raven.Server.Documents.AI.Embeddings;
+
+public class QueryEmbeddingsCacher : BackgroundWorkBase
+{
+    private readonly DocumentDatabase _database;
+
+    private readonly ConcurrentQueue<EmbeddingGenerationItem> _embeddingsQueue;
+    private readonly AsyncManualResetEvent _mre;
+    private int _approximateCount;
+    public QueryEmbeddingsCacher(DocumentDatabase database, CancellationToken shutdown) : base(database.Name, database.Loggers.GetLogger<QueryEmbeddingsCacher>(), shutdown)
+    {
+        _database = database;
+        _embeddingsQueue = new ConcurrentQueue<EmbeddingGenerationItem>();
+        _mre = new AsyncManualResetEvent();
+    }
+
+    protected override async Task DoWork()
+    {
+        while (CancellationToken.IsCancellationRequested == false)
+        {
+            await _mre.WaitAsync(CancellationToken);
+
+            _mre.Reset();
+
+            var mightBeMore = CacheEnqueuedEmbeddings();
+            
+            if (mightBeMore)
+                _mre.Set();
+        }
+    }
+
+    internal bool CacheEnqueuedEmbeddings()
+    {
+        var mightBeMore = false;
+        
+        var maxBatchSize = _database.Configuration.Ai.QueryEmbeddingsGenerationMaxCacheBatchSize;
+
+        var payload = new List<EmbeddingGenerationItem>(Math.Min(_approximateCount, maxBatchSize));
+
+        while (_embeddingsQueue.TryDequeue(out var item))
+        {
+            CancellationToken.ThrowIfCancellationRequested();
+                
+            Interlocked.Decrement(ref _approximateCount);
+
+            payload.Add(item);
+
+            if (payload.Count >= maxBatchSize)
+            {
+                mightBeMore = true;
+                break;
+            }
+        }
+
+        if (payload.Count == 0)
+            return mightBeMore;
+
+        var putEmbeddingsCommand = new PutQueryEmbeddingsCommand(payload, _database);
+
+        _database.TxMerger.EnqueueSync(putEmbeddingsCommand);
+
+        return mightBeMore;
+    }
+
+    public void EnqueueEmbeddingsToCache(List<EmbeddingGenerationItem> embeddings)
```

Comment:
What is the life cycle of the values here? Are we just talking about managed objects?
What about buffering?

Reply:
This is used by querying code. The embedding values come from query parameters (`GetEmbeddingsForQueryAsync`). Yes, we're talking about managed objects (in particular the embeddings are `ReadOnlyMemory<float>`).

Not sure that I follow the question about buffering.